File: Navrhove vzory/cast2/sources/observer.cs
abstract class ASubject
{
    //This is the one way we can implement it, lets call it WAY_1
    ArrayList list = new ArrayList();

    //This is another way we can implement observer, lets call it WAY_2
    public delegate void StatusUpdate(float price);
    public event StatusUpdate OnStatusUpdate = null;

    public void Attach(Shop product)
    {
        //For way 1 lets assign attach the observers with subjects
        list.Add(product);
    }
    public void Detach(Shop product)
    {
        //For way 1 lets assign detach the observers with subjects
        list.Remove(product);
    }

    public void Attach2(Shop product)
    {
        //For way 2 lets assign attach the observers with subjects
        OnStatusUpdate += new StatusUpdate(product.Update);
    }
    public void Detach2(Shop product)
    {
        //For way 2 lets assign detach the observers with subjects
        OnStatusUpdate -= new StatusUpdate(product.Update);
    }

    public void Notify(float price)
    {
        //For way 1 lets notify the observers with change
        foreach (Shop p in list)
        {
            p.Update(price);
        }

        //For way 2 lets notify the observers with change
        if (OnStatusUpdate != null)
        {
            OnStatusUpdate(price);
        }
    }
}



class DummyProduct : ASubject
{
    public void ChangePrice(float price)
    {
        Notify(price);
    }
}


interface IObserver
{
    void Update(float price);
}


class Shop : IObserver
{
    //Name if the product
    string name;
    float price = 0.0f; //default

    public Shop(string name)
    {
        this.name = name;
    }
    #region IObserver Members

    public void Update(float price)
    {
        this.price = price;

        //Lets print on console just to test the working
        Console.WriteLine(@"Price at {0} is now {1}", name, price);
    }

    #endregion
}



static void Main(string[] args)
{
    DummyProduct product = new DummyProduct();

    // We have four shops wanting to keep updated price set by product owner
    Shop shop1 = new Shop("Shop 1");
    Shop shop2 = new Shop("Shop 2");

    Shop shop3 = new Shop("Shop 3");
    Shop shop4 = new Shop("Shop 4");

    //Lets use WAY_1 for first two shops
    product.Attach(shop1);
    product.Attach(shop2);

    //Lets use WAY_2 for other two shops
    product.Attach2(shop3);
    product.Attach2(shop4);

    //Now lets try changing the products price, this should update the shops automatically
    product.ChangePrice(23.0f);

    //Now shop2 and shop 4 are not interested in new prices so they unsubscribe
    product.Detach(shop2);
    product.Detach2(shop4);

    //Now lets try changing the products price again
    product.ChangePrice(26.0f);

    Console.Read();
}

File: Navrhove vzory/cast2/sources/observer.cpp
//Header File
#pragma once
#include <vector>
#include <list>
#include "shop.h"

class ASubject
{
    //Lets keep a track of all the shops we have observing
    std::vector<Shop*> list;

public:
    void Attach(Shop *product);
    void Detach(Shop *product);
    void Notify(float price); 
};

//CPP File
#include "ASubject.h"
#include <algorithm>

using namespace std;

void ASubject::Attach(Shop *shop)
{
    list.push_back(shop);
}
void ASubject::Detach(Shop *shop)
{    
    list.erase(std::remove(list.begin(), list.end(), shop), list.end());    
}

void ASubject::Notify(float price)
{
    for(vector<Shop*>::const_iterator iter = list.begin(); iter != list.end(); ++iter)
    {
        if(*iter != 0)
        {
            (*iter)->Update(price);
        }
    }
}


//Header File
#pragma once
#include "ASubject.h"

class DummyProduct : public ASubject
{
public:
    void ChangePrice(float price);
};

//CPP File
#include "DummyProduct.h"

void DummyProduct::ChangePrice(float price)
{
    Notify(price);
}

#pragma once

class IObserver
{
public:
    virtual void Update(float price) = 0;
};


//Header File
#pragma once
#include <iostream>
#include <string>
#include "IObserver.h"

class Shop : IObserver
{
    //Name of the Shop
    std::string name;
    float price;
public:
    Shop(std::string n); 
    void Update(float price);          
};

//CPP File
#include "Shop.h"

Shop::Shop(std::string name)
{
    this->name = name;
}

void Shop::Update(float price)
{
    this->price = price;

    //Lets print on console just to test the working
    std::cout << "Price at "<< name << " is now "<< price << "\n";
}



int main(int argc, char* argv[])
{
    DummyProduct product;
                    
    // We have four shops wanting to keep updated price set by product owner
    Shop shop1("Shop 1");
    Shop shop2("Shop 2");

    product.Attach(&shop1);
    product.Attach(&shop2);

    //Now lets try changing the products price, this should update the shops automatically
    product.ChangePrice(23.0f);

    //Now shop2 is not interested in new prices so they unsubscribe
    product.Detach(&shop2);            

    //Now lets try changing the products price again
    product.ChangePrice(26.0f);

    getchar();
    return 0;
}